framework/DataObjects/XDataSet.py

```diff
@@ -501,6 +501,29 @@ def reset(self):
     self._meta = {}
     # TODO others?
 
+  def sliceByIndex(self,axis):
+    """
+      Returns list of realizations at "snapshots" along "axis".
+      For example, if axis is 'time', then returns cross-sectional slices of the dataobject at each recorded 'time' index value.
+      @ In, axis, str, name of index along which to obtain slices
+      @ Out, slices, list, list of xr.Dataset slices.
+    """
+    data = self.asDataset()
+    # if empty, nothing to do
+    if self._data is None or len(self._data) == 0:
+      self.raiseAWarning('Tried to return sliced data, but DataObject is empty!')
+      return []
+    # assert that axis is an index
+    if axis not in self.indexes + [self.sampleTag]:
+      self.raiseAnError(IOError,'Requested slices along "{}" but that variable is not an index!  Options are: {}'.format(axis,self.indexes))
+    numAxisValues = len(data[axis])
+    # TODO potentially slow loop
+    slices = [None]*numAxisValues
+    for i in range(numAxisValues):
+      slices[i] = data.isel(**{axis:i})
+      # NOTE: The slice may include NaN if a variable does not have a value along a different index for this snapshot along "axis"
```

Comment:
can you do a vectoring index instead of looping?

Reply:
done